File: C/doubly-linked.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct Node {
    int data;
    struct Node *prev, *next;
} Node;

Node *head = NULL;

void insertAtBeginning(int data) {
    Node *newNode = (Node *) malloc(sizeof(Node));
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = head;

    if (head != NULL) {
        head->prev = newNode;
    }

    head = newNode;
}

void insertAtEnd(int data) {
    Node *newNode = (Node *) malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = NULL;

    if (head == NULL) {
        newNode->prev = NULL;
        head = newNode;
        return;
    }

    Node *temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }

    temp->next = newNode;
    newNode->prev = temp;
}

void deleteAtBeginning() {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }

    Node *temp = head;

    if (head->next == NULL) {
        head = NULL;
        free(temp);
        return;
    }

    head = head->next;
    head->prev = NULL;
    free(temp);
}

void deleteAtEnd() {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }

    Node *temp = head;

    if (head->next == NULL) {
        head = NULL;
        free(temp);
        return;
    }

    while (temp->next != NULL) {
        temp = temp->next;
    }

    temp->prev->next = NULL;
    free(temp);
}

void displayList() {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }

    Node *temp = head;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }

    printf("\n");
}

int main() {
    srand(time(NULL));

    int n;
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    for (int i = 0; i < n; i++) {
        int data = rand() % 100;
        insertAtEnd(data);
    }

    printf("Original doubly linked list: ");
    displayList();

    int choice;
    do {
        printf("\n1. Insert at beginning\n");
        printf("2. Insert at end\n");
        printf("3. Delete at beginning\n");
        printf("4. Delete at end\n");
        printf("5. Display list\n");
        printf("6. Exit\n");

        printf("\nEnter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                int data;
                printf("Enter data to insert: ");
                scanf("%d", &data);
                insertAtBeginning(data);
                break;
            case 2:
                data;
                printf("Enter data to insert: ");
                scanf("%d", &data);
                insertAtEnd(data);
                break;
            case 3:
                deleteAtBeginning();
                break;
            case 4:
                deleteAtEnd();
                break;
            case 5:
                displayList();
                break;
            case 6:
                printf("Exiting...\n");
                break;
            default:
                printf("Invalid choice!\n");
        }
    } while (choice != 6);

    return 0;
}

File: Python/linkedList/doublyLinked.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_at_beginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        if self.head is not None:
            self.head.prev = new_node
        self.head = new_node
        if self.tail is None:
            self.tail = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if self.tail is not None:
            self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node
        if self.head is None:
            self.head = new_node

    def insert_after_node(self, data, after_node):
        if after_node is None:
            raise ValueError("Invalid node")

        new_node = Node(data)
        new_node.next = after_node.next
        if after_node.next is not None:
            after_node.next.prev = new_node
        after_node.next = new_node
        new_node.prev = after_node

    def delete_at_beginning(self):
        if self.head is None:
            return None

        deleted_node = self.head
        self.head = self.head.next
        if self.head is not None:
            self.head.prev = None
        deleted_node.next = None
        deleted_node.prev = None

        if self.head is None:
            self.tail = None

        return deleted_node.data

    def delete_at_end(self):
        if self.tail is None:
            return None

        deleted_node = self.tail
        self.tail = self.tail.prev
        if self.tail is not None:
            self.tail.next = None
        deleted_node.next = None
        deleted_node.prev = None

        if self.head is None:
            self.tail = None

        return deleted_node.data

    def delete_node(self, node):
        if node is None:
            raise ValueError("Invalid node")

        if node is self.head:
            self.delete_at_beginning()
            return

        if node is self.tail:
            self.delete_at_end()
            return

        node.prev.next = node.next
        node.next.prev = node.prev
        node.next = None
        node.prev = None

    def display(self):
        current_node = self.head
        while current_node is not None:
            print(current_node.data)
            current_node = current_node.next

    def reverse(self):
        current_node = self.head
        while current_node is not None:
            temp = current_node.next
            current_node.next = current_node.prev
            current_node.prev = temp
            current_node = temp

        temp = self.head
        self.head = self.tail
        self.tail = temp
